test/wt_hang_analyzer/wt_hang_analyzer.py

```diff
@@ -0,0 +1,608 @@
+#!/usr/bin/env python
+"""Hang Analyzer module.
+
+A prototype hang analyzer for Evergreen integration to help investigate test timeouts
+
+1. Script supports taking dumps, and/or dumping a summary of useful information about a process
+2. Script will iterate through a list of interesting processes,
+    and run the tools from step 1. The list of processes can be provided as an option.
+
+Currenty only supports Linux. There are two issues with the MacOS and windows implementation:
+1. lldb cannot attach to processes in MacOS.
+2. Windows cannot find the debug symbols.
+"""
+
+import csv
+import glob
+import itertools
+import logging
+import os
+import platform
+import re
+import signal
+import subprocess
+import sys
+import tempfile
+import threading
+import traceback
+import time
+from distutils import spawn
+from io import BytesIO, TextIOWrapper
+from optparse import OptionParser
+_IS_WINDOWS = (sys.platform == "win32")
+
+if _IS_WINDOWS:
+    import win32event
+    import win32api
+
+"""
+Helper class to read output of a subprocess.
+
+Used to avoid deadlocks from the pipe buffer filling up and blocking the subprocess while it's
+being waited on.
+"""
+class LoggerPipe(threading.Thread):
+    """Asynchronously reads the output of a subprocess and sends it to a logger."""
+
+    # The start() and join() methods are not intended to be called directly on the LoggerPipe
+    # instance. Since we override them for that effect, the super's version are preserved here.
+    __start = threading.Thread.start
+    __join = threading.Thread.join
+
+    def __init__(self, logger, level, pipe_out):
+        """Initialize the LoggerPipe with the specified arguments."""
+
+        threading.Thread.__init__(self)
+        # Main thread should not call join() when exiting
+        self.daemon = True
+
+        self.__logger = logger
+        self.__level = level
+        self.__pipe_out = pipe_out
+
+        self.__lock = threading.Lock()
+        self.__condition = threading.Condition(self.__lock)
+
+        self.__started = False
+        self.__finished = False
+
+        LoggerPipe.__start(self)
+
+    def start(self):
+        """Start not implemented."""
+        raise NotImplementedError("start should not be called directly")
+
+    def run(self):
+        """Read the output from 'pipe_out' and logs each line to 'logger'."""
+
+        with self.__lock:
+            self.__started = True
+            self.__condition.notify_all()
+
+        # Close the pipe when finished reading all of the output.
```

Comment:
This needs an additional word or two, or to be rearranged. I suggest "Close the pipe when we have finished reading all of the output." or "Close the pipe when all of the output has been read."

Reply:
Thanks, addressed in ef39187